source/link/linker.cpp

```diff
@@ -0,0 +1,755 @@
+            break;
+          case SpvOpTypeStruct:
+          case SpvOpTypeFunction:
+            for (uint32_t j = 0u; j < typesIter->NumInOperands(); ++j)
+              i.second.emplace(typesIter->GetSingleWordInOperand(j));
+            break;
+          case SpvOpTypePointer:
+            i.second.emplace(typesIter->GetSingleWordInOperand(1u));
+            break;
+          default:
+            break;
+        }
+      }
+    } while (typesIter != modules[i.first]->types_values_begin());
+  }
+
+  // Ensure the import and export types are similar
+  for (const auto&i : linkingsToDo) {
+    auto importTypeIter = types.find(i.first.typeId);
+    assert(importTypeIter != types.end());
+    auto exportTypeIter = types.find(i.second.typeId);
+    assert(exportTypeIter != types.end());
+
+    if (!areTypesSimilar(importTypeIter->second.i, exportTypeIter->second.i, types))
+      return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                        SPV_ERROR_INVALID_BINARY)
+             << "Type mismatch between imported variable/function %" << i.first.id
+             << " and exported variable/function %" << i.second.id << ".";
+  }
+
+
+  // Ensure the import and export decorations are similar
+  for (const auto& i : linkingsToDo) {
+    if (!haveIdsSimilarDecorations(i.first.id, modules[i.first.moduleIndex],
+                                   i.second.id, modules[i.second.moduleIndex]))
+        return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                          SPV_ERROR_INVALID_BINARY)
+               << "Decorations mismatch between imported variable/function %" << i.first.id
+               << " and exported variable/function %" << i.second.id << ".";
+    if (!haveIdsSimilarDecorations(i.first.typeId, modules[i.first.moduleIndex],
+                                   i.second.typeId, modules[i.second.moduleIndex]))
+        return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                          SPV_ERROR_INVALID_BINARY)
+               << "Decorations mismatch between the type of imported variable/function %" << i.first.id
+               << " and the type of exported variable/function %" << i.second.id << ".";
+    for (uint32_t j = 0u; j < i.first.parametersIds.size(); ++j)
+      if (!haveIdsSimilarDecorations(i.first.parametersIds[j], modules[i.first.moduleIndex],
+                                     i.second.parametersIds[j], modules[i.second.moduleIndex]))
+          return libspirv::DiagnosticStream(position, impl_->context->consumer,
+                                            SPV_ERROR_INVALID_BINARY)
+                 << "Decorations mismatch between imported function %" << i.first.id << "'s"
+                 << " and exported function %" << i.second.id << "'s " << j << "th parameter.";
+  }
+
+  // TODO(pierremoreau): Remove import types
```

Comment:
I think you just delegate this to the RemoveDuplicatesPass.  But you should say so here anyway.